src/couch_replicator_connection.erl

```diff
@@ -0,0 +1,179 @@
+    timer
+}).
+
+-record(connection, {
+    worker,
+    host,
+    port,
+    owner
+}).
+
+-include_lib("ibrowse/include/ibrowse.hrl").
+
+
+start_link() ->
+    gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).
+
+
+init([]) ->
+    process_flag(trap_exit, true),
+    ?MODULE = ets:new(?MODULE, [named_table, public, {keypos, #connection.worker}]),
+    ok = config:listen_for_changes(?MODULE, self()),
+    Interval = config:get_integer("replicator", "connection_close_interval", ?DEFAULT_CLOSE_INTERVAL),
+    {ok, Timer} = timer:send_after(Interval, close_idle_connections),
+    {ok, #state{close_interval=Interval, timer=Timer}}.
+
+
+acquire(URL) ->
+    case gen_server:call(?MODULE, {acquire, URL}) of
+        {ok, Worker} ->
+            link(Worker),
+            {ok, Worker};
+        {error, all_allocated} ->
+            {ok, Pid} = ibrowse:spawn_link_worker_process(URL),
+            ok = gen_server:call(?MODULE, {create, URL, Pid}),
+            {ok, Pid};
+        {error, Reason} ->
+            {error, Reason}
+    end.
+
+
+relinquish(Worker) ->
+    unlink(Worker),
+    gen_server:cast(?MODULE, {relinquish, Worker}).
+
+
+handle_call({acquire, URL}, From, State) ->
+    {Pid, _Ref} = From,
+    monitor(process, Pid),
+    case ibrowse_lib:parse_url(URL) of
+        #url{host=Host, port=Port} ->
+            case ets:match_object(?MODULE, #connection{host=Host, port=Port, owner=undefined, _='_'}, 1) of
+                '$end_of_table' ->
+                    {reply, {error, all_allocated}, State};
+                {[Worker], _Cont} ->
+                    couch_stats:increment_counter([couch_replicator, connection, acquires]),
+                    ets:insert(?MODULE, Worker#connection{owner=Pid}),
+                    {reply, {ok, Worker#connection.worker}, State}
+            end
+    end;
+
+handle_call({create, URL, Worker}, From, State) ->
+    {Pid, _Ref} = From,
+    case ibrowse_lib:parse_url(URL) of
+        #url{host=Host, port=Port} ->
+            link(Worker),
+            couch_stats:increment_counter([couch_replicator, connection, creates]),
+            true = ets:insert_new(
+                ?MODULE,
+                #connection{host=Host, port=Port, owner=Pid, worker=Worker}
+            ),
+            {reply, ok, State}
+    end.
+
+
+handle_cast({relinquish, WorkerPid}, State) ->
+    couch_stats:increment_counter([couch_replicator, connection, relinquishs]),
+    case ets:lookup(?MODULE, WorkerPid) of
+        [Worker] ->
+            ets:insert(?MODULE, Worker#connection{owner=undefined});
+        [] ->
+            ok
+    end,
+    {noreply, State}.
+
+
+% owner crashed
+handle_info({'DOWN', _Ref, process, Pid, _Reason}, State) ->
+    couch_stats:increment_counter([couch_replicator, connection, owner_crashes]),
+    Workers = ets:match_object(?MODULE, #connection{owner=Pid, _='_'}),
+    lists:foreach(fun(Worker) ->
+        ets:insert(?MODULE, Worker#connection{owner=undefined})
```

Comment:
Thinking about it some more. Monitoring is necessary because owner could exit `normal` and int that case workers won't die (they are linked to owner and if they don't trap exit then they won't notice owner died).